### ISSUE TITLE
Consider caching parsing result

### ISSUE THREAD

**Issue by Pamplemousse** [semgrep/semgrep#4940]
**Is your feature request related to a problem? Please describe.**

On a big codebase, a lot of time is spent parsing. It makes it painful to use `semgrep` for quick searches, and prototyping rules.

```shell
Semgrep timing summary:
Total time: 53.1036 Config time: 0.0731 Core time: 53.0153 Ignores time: 0.0146

Semgrep-core time:
Total CPU time: 37.1760  File parse time: 25.1105  Rule parse time: 0.0161  Match time: 0.3193
Slowest 5/51669 files
[...] (352KB): 11.5239
[...] ( 63KB): 1.3864
[...] (244KB): 1.3027
[...] ( 65KB): 1.2080
[...] (311KB): 1.1632
Slowest 5 rules to run (excluding parse time)
test-validation:                                                 run time 12.0655  match time 0.3193
ran 1 rules on 51669 files: 0 findings
```

**Describe the solution you'd like**

It would be awesome if there were some caching of the AST once the parsing is done. Cache keys should be the hashes of files' content.

Caching should be off by default for compatibility with current behaviour.

**Describe alternatives you've considered**

Scoping down the search to specific folders in the codebase.
Con: it requires to know where to find issues before searching.

**Use case**

Faster feedback loop when writing rules at the big codebase level.

**Comment by Pamplemousse** [comment]
Interestingly, excluding the file that ate 11s seems to reduce parsing by the same amount, but only reduces the overall time by 1s; Not sure parsing is the only thing to blame for the slowness here...

```shell
Semgrep timing summary:
Total time: 52.3292 Config time: 0.0725 Core time: 52.2414 Ignores time: 0.0146

Semgrep-core time:
Total CPU time: 24.9688  File parse time: 12.9667  Rule parse time: 0.0117  Match time: 0.2677
Slowest 5/51668 files
[...] ( 63KB): 1.2985
[...] (244KB): 1.1907
[...] (311KB): 1.1340
[...] ( 65KB): 1.1337
[...] (271KB): 0.9226
Slowest 5 rules to run (excluding parse time)
test-validation:                                                 run time 12.0022  match time 0.2677
ran 1 rules on 51668 files: 0 findings
```

**Comment by r2c-demo** [comment]
This issue is synced in Linear at https://linear.app/r2c/issue/PA-1087/consider-caching-parsing-result.

**Comment by aryx** [comment]
cc @emjin

**Comment by emjin** [comment]
Just to help you understand why the numbers don't quite add up: semgrep-core is run in parallel, but the parse vs match time is reported as if it were run on one core. So even though the semgrep-core total CPU time is 24 seconds, that might only be 8 seconds in practice.

**Comment by emjin** [comment]
What's probably taking a long time is getting the files, which we've noticed is slow. We definitely want to optimize that, but it's not a priority right now. A workaround you can use, if all you're changing is the content of your rules, is to run 

`semgrep --config [your_config] . --dump-command-for-core`

This will output the semgrep-core command semgrep runs. You can change the content of your rules and run that command instead to skip the overhead + file targeting, and it should work. No guarantees about the quality of the UX, but this might help you out before we have time to focus on our CLI performance

I'm also curious, if you do run this, how long it takes to output the semgrep-core command and whether it's indeed taking most of the time.

**Comment by aryx** [comment]
This is solved by https://github.com/returntocorp/semgrep/pull/7744
